sklearn/metrics/pairwise.py

```diff
@@ -585,6 +585,189 @@ def cosine_distances(X, Y=None):
+
+    # Calculates the normalized ranges and max values of numeric values
+    _, num_cols = X_num.shape
+    ranges_of_numeric = np.zeros(num_cols)
+    max_of_numeric = np.zeros(num_cols)
+    for col in range(num_cols):
+        col_array = X_num[:, col].astype(np.float32)
+        max = np.nanmax(col_array)
+        min = np.nanmin(col_array)
+
+        if np.isnan(max):
+            max = 0.0
+        if np.isnan(min):
+            min = 0.0
+        max_of_numeric[col] = max
+        ranges_of_numeric[col] = (1 - min / max) if (max != 0) else 0.0
+
+    # This is to normalize the numeric values between 0 and 1.
+    X_num = np.divide(X_num, max_of_numeric, out=np.zeros_like(X_num),
+                      where=max_of_numeric != 0)
+
+    if feature_weight is None:
+        feature_weight = np.ones(n_cols)
+
+    feature_weight_cat = feature_weight[categorical_features]
+    feature_weight_num = feature_weight[np.logical_not(categorical_features)]
+
+    y_n_rows, _ = Y.shape
+
+    dm = np.zeros((n_rows, y_n_rows), dtype=np.float32)
+
+    feature_weight_sum = feature_weight.sum()
+
+    Y_cat = None
+    Y_num = None
+
+    if not y_none:
+        Y_cat = Y[:, categorical_features]
+        Y_num = Y[:, np.logical_not(categorical_features)]
+        # This is to normalize the numeric values between 0 and 1.
+        Y_num = np.divide(Y_num, max_of_numeric, out=np.zeros_like(Y_num),
+                          where=max_of_numeric != 0)
+    else:
+        Y_cat = X_cat
+        Y_num = X_num
+
+    for i in range(n_rows):
+        j_start = i
+
+        # for non square results
+        if n_rows != y_n_rows:
+            j_start = 0
+
+        Y_cat[j_start:n_rows, :]
+        Y_num[j_start:n_rows, :]
+        result = _gower_distance_row(X_cat[i, :], X_num[i, :],
+                                     Y_cat[j_start:n_rows, :],
+                                     Y_num[j_start:n_rows, :],
+                                     feature_weight_cat,
+                                     feature_weight_num,
+                                     feature_weight_sum,
+                                     categorical_features,
+                                     ranges_of_numeric,
+                                     max_of_numeric)
+        dm[i, j_start:] = result
+        dm[i:, j_start] = result
+
+    return dm
+
+
+def _gower_distance_row(xi_cat, xi_num, xj_cat, xj_num, feature_weight_cat,
+                        feature_weight_num, feature_weight_sum,
+                        categorical_features, ranges_of_numeric,
+                        max_of_numeric):
+
+    """Computes the gower distances between a set of rows of X and Y
+    """
+    # categorical columns
+    sij_cat = np.where(xi_cat == xj_cat, np.zeros_like(xi_cat),
+                       np.ones_like(xi_cat))
+
+    sum_cat = np.multiply(feature_weight_cat, sij_cat).sum(axis=1)
+
+    # numerical columns
+    abs_delta = np.absolute(xi_num-xj_num)
```

Comment:
Space around - please

Reply:
Do you deal with NaN somewhere?

Reply:
>> Do you deal with NaN somewhere? 
A: NaN operations and NaN non proliferation are backed along this code. Unit tests are provided to support this.